protomap/proto.go

```diff
@@ -700,6 +726,137 @@ func protoFromPathsInternal(p proto.Message, vals map[*gpb.Path]any, valPrefix,
+		for _, s := range seen {
+			if reflect.DeepEqual(s, key) {
+				done = true
+				break
+			}
+		}
+		if done {
+			continue
+		}
+		seen = append(seen, key)
+
+		listElemChildren, err := findChildren(vals, valPrefix, keyPaths[i], false, false)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("logic error, error returned from extracting list member children, %v", err)
+		}
+
+		// We now need to create the "XXXKey" message, and populate the key values, subsequent values are then populated
+		// into the one protobuf message field.
+		childMsgEmpty := le.NewElement().Message()
+		childMsgTarget := le.NewElement().Message()
+
+		// Walk through the fields of the XXXKey message that we just created. We use the childMsgEmpty here so that we
+		// don't change the message whilst iterating which causes us to revisit that field. We set the values in
+		// the childMsgTarget message.
+		var retErr error
+		unpopRange{childMsgEmpty}.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+			// We have one field that is a message in a key message, which is the payload. The remaining fields are the keys.
+			switch fd.Kind() {
+			case protoreflect.MessageKind:
+				m := childMsgTarget.NewField(fd).Message()
+				// We must ignore extra fields from this point in the recursion, because keys map to fields that
+				// are not present in the generated protobuf.
+				if err := protoFromPathsInternal(m.Interface(), listElemChildren, keyPaths[i], keyPaths[i], true); err != nil {
+					retErr = err
+					return false
+				}
+				childMsgTarget.Set(fd, protoreflect.ValueOfMessage(m))
+			default:
+				paths, err := annotatedSchemaPath(fd)
+				if err != nil {
+					retErr = err
+					return false
+				}
+
+				var keyName string
+				for _, p := range paths {
+					keyName = p.Elem[len(p.Elem)-1].Name
+					break
+				}
+				pv, err := listKeyAsProtoValue(fd, key[keyName])
+				if err != nil {
+					retErr = fmt.Errorf("field %s, %v", fd.FullName(), err)
+					return false
+				}
+				childMsgTarget.Set(fd, pv)
+			}
+
+			return true
+		})
+		if retErr != nil {
+			return protoreflect.Value{}, fmt.Errorf("field %s, %v", fd.FullName(), retErr)
+		}
+		le.Append(protoreflect.ValueOfMessage(childMsgTarget))
+	}
+
+	return protoreflect.ValueOfList(le), nil
+}
+
+// listKeyAsProtoValue converts the value of a list key (represented as a string) into a protoreflect.Value that can be
+// used to set a scalar protobuf field.
+func listKeyAsProtoValue(fd protoreflect.FieldDescriptor, val string) (protoreflect.Value, error) {
+	switch fd.Kind() {
+	case protoreflect.Uint64Kind:
+		v, err := strconv.ParseUint(val, 10, 64)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("invalid uint64 value %v, err: %v", val, err)
+		}
+		return protoreflect.ValueOfUint64(v), nil
+	case protoreflect.StringKind:
+		return protoreflect.ValueOfString(val), nil
+	default:
+		return protoreflect.Value{}, fmt.Errorf("invalid kind %v", fd.Kind())
```

Comment:
Could there be other types? If so change the error message to say "unsupported"?

Reply:
Done.